include/deal.II/distributed/fully_distributed_tria.h

```diff
@@ -385,6 +401,80 @@ namespace parallel
         currently_processing_prepare_coarsening_and_refinement_for_internal_usage;
     };
 
+
+#ifndef DOXYGEN
+
+    namespace
+    {
+      template <class Archive>
+      void serialize(CellData<3> &cell,
```

Comment:
Why do you need these specializations? Does it not work if you did
```
  if (dim >= 3) {
    ...serialize those members that exist only in 3d...
  }
```
?

Reply:
No it does not. The boost functions are templated and the if-statements are evaluated just later during compiling...

Reply:
I'm not sure we're talking about the same thing. What I mean is this: Why doesn't an implementation like the following for the member function `CellData::serialize(...)` work:
```
void CellData::serialize(...)
{
        ar &cell.id;
        ar &cell.subdomain_id;
        ar &cell.level_subdomain_id;
        ar &cell.manifold_id;
        if (dim>=2)
          ar &cell.manifold_line_ids;
        if (dim>=3)
          ar &cell.manifold_quad_ids;
        ar &cell.boundary_ids;
}
```
This would avoid the need for the three overloaded functions in the anonymous namespace.

Reply:
```cpp
4676: Test command: /lnm/bin/cmake "-DTRGT=parallel_fullydistributed_construction_data_1.debug.diff" "-DTEST=serialization/parallel_fullydistributed_construction_data_1.debug" "-DEXPECT=PASSED" "-DBINARY_DIR=/home/munch/sw-index/dealii-build/tests/serialization" "-DGUARD_
FILE=/home/munch/sw-index/dealii-build/tests/serialization/parallel_fullydistributed_construction_data_1.debug/interrupt_guard.cc" "-P" "/home/munch/sw-index/dealii-build/share/deal.II/scripts/run_test.cmake"
4676: Test timeout computed to be: 600
4676: In file included from /home/munch/sw-index/dealii/bundled/boost-1.70.0/include/boost/serialization/split_member.hpp:23,
4676:                  from /home/munch/sw-index/dealii/bundled/boost-1.70.0/include/boost/serialization/nvp.hpp:26,
4676:                  from /home/munch/sw-index/dealii/bundled/boost-1.70.0/include/boost/serialization/array_wrapper.hpp:19,
4676:                  from /home/munch/sw-index/dealii/bundled/boost-1.70.0/include/boost/archive/basic_binary_iprimitive.hpp:53,
4676:                  from /home/munch/sw-index/dealii/bundled/boost-1.70.0/include/boost/archive/binary_iarchive_impl.hpp:20,
4676:                  from /home/munch/sw-index/dealii/bundled/boost-1.70.0/include/boost/archive/binary_iarchive.hpp:20,
4676:                  from /home/munch/sw-index/dealii/include/deal.II/base/utilities.h:45,
4676:                  from /home/munch/sw-index/dealii/include/deal.II/base/tensor.h:27,
4676:                  from /home/munch/sw-index/dealii/include/deal.II/base/symmetric_tensor.h:23,
4676:                  from /home/munch/sw-index/dealii/include/deal.II/base/array_view.h:23,
4676:                  from /home/munch/sw-index/dealii/include/deal.II/base/mpi.h:21,
4676:                  from /home/munch/sw-index/dealii/include/deal.II/distributed/tria_base.h:22,
4676:                  from /home/munch/sw-index/dealii/include/deal.II/distributed/fully_distributed_tria.h:22,
4676:                  from /home/munch/sw-index/dealii/tests/serialization/parallel_fullydistributed_construction_data_1.cc:19:
4676: /home/munch/sw-index/dealii/bundled/boost-1.70.0/include/boost/serialization/access.hpp: In instantiation of ‘static void boost::serialization::access::serialize(Archive&, T&, unsigned int) [with Archive = boost::archive::text_oarchive; T = unsigned int [0]]’:
4676: /home/munch/sw-index/dealii/bundled/boost-1.70.0/include/boost/serialization/serialization.hpp:68:22:   required from ‘void boost::serialization::serialize(Archive&, T&, unsigned int) [with Archive = boost::archive::text_oarchive; T = unsigned int [0]]’
4676: /home/munch/sw-index/dealii/bundled/boost-1.70.0/include/boost/serialization/serialization.hpp:126:14:   required from ‘void boost::serialization::serialize_adl(Archive&, T&, unsigned int) [with Archive = boost::archive::text_oarchive; T = unsigned int [0]]’
4676: /home/munch/sw-index/dealii/bundled/boost-1.70.0/include/boost/archive/detail/oserializer.hpp:153:40:   required from ‘void boost::archive::detail::oserializer<Archive, T>::save_object_data(boost::archive::detail::basic_oarchive&, const void*) const [with Archive = 
boost::archive::text_oarchive; T = unsigned int [0]]’
4676: /home/munch/sw-index/dealii/bundled/boost-1.70.0/include/boost/archive/detail/oserializer.hpp:106:1:   required from ‘class boost::archive::detail::oserializer<boost::archive::text_oarchive, unsigned int [0]>’
4676: /home/munch/sw-index/dealii/bundled/boost-1.70.0/include/boost/archive/detail/oserializer.hpp:258:13:   required from ‘static void boost::archive::detail::save_non_pointer_type<Archive>::save_standard::invoke(Archive&, const T&) [with T = unsigned int [0]; Archive =
 boost::archive::text_oarchive]’
4676: /home/munch/sw-index/dealii/bundled/boost-1.70.0/include/boost/archive/detail/oserializer.hpp:275:38:   [ skipping 115 instantiation contexts, use -ftemplate-backtrace-limit=0 to disable ]
4676: /home/munch/sw-index/dealii/bundled/boost-1.70.0/include/boost/archive/detail/common_oarchive.hpp:71:22:   required from ‘void boost::archive::detail::common_oarchive<Archive>::save_override(T&) [with T = const dealii::parallel::fullydistributed::ConstructionData<2,
 2>; Archive = boost::archive::text_oarchive]’
4676: /home/munch/sw-index/dealii/bundled/boost-1.70.0/include/boost/archive/basic_text_oarchive.hpp:83:9:   required from ‘void boost::archive::basic_text_oarchive<Archive>::save_override(T&) [with T = const dealii::parallel::fullydistributed::ConstructionData<2, 2>; Arc
hive = boost::archive::text_oarchive]’
4676: /home/munch/sw-index/dealii/bundled/boost-1.70.0/include/boost/archive/detail/interface_oarchive.hpp:70:9:   required from ‘Archive& boost::archive::detail::interface_oarchive<Archive>::operator<<(const T&) [with T = dealii::parallel::fullydistributed::ConstructionD
ata<2, 2>; Archive = boost::archive::text_oarchive]’
4676: /home/munch/sw-index/dealii/tests/serialization/serialization.h:55:8:   required from ‘void verify(const T&, T&) [with T = dealii::parallel::fullydistributed::ConstructionData<2, 2>]’
4676: /home/munch/sw-index/dealii/tests/serialization/parallel_fullydistributed_construction_data_1.cc:51:9:   required from ‘void test(MPI_Comm) [with int dim = 2; MPI_Comm = ompi_communicator_t*]’
4676: /home/munch/sw-index/dealii/tests/serialization/parallel_fullydistributed_construction_data_1.cc:75:15:   required from here
4676: /home/munch/sw-index/dealii/bundled/boost-1.70.0/include/boost/serialization/access.hpp:116:11: error: request for member ‘serialize’ in ‘t’, which is of non-class type ‘unsigned int [0]’
4676:          t.serialize(ar, file_version);
4676:          ~~^~~~~~~~~
4676: In file included from /home/munch/sw-index/dealii/bundled/boost-1.70.0/include/boost/integer.hpp:23,
4676:                  from /home/munch/sw-index/dealii/bundled/boost-1.70.0/include/boost/archive/basic_binary_iprimitive.hpp:48,
4676:                  from /home/munch/sw-index/dealii/bundled/boost-1.70.0/include/boost/archive/binary_iarchive_impl.hpp:20,
4676:                  from /home/munch/sw-index/dealii/bundled/boost-1.70.0/include/boost/archive/binary_iarchive.hpp:20,
4676:                  from /home/munch/sw-index/dealii/include/deal.II/base/utilities.h:45,
4676:                  from /home/munch/sw-index/dealii/include/deal.II/base/tensor.h:27,
4676:                  from /home/munch/sw-index/dealii/include/deal.II/base/symmetric_tensor.h:23,
4676:                  from /home/munch/sw-index/dealii/include/deal.II/base/array_view.h:23,
4676:                  from /home/munch/sw-index/dealii/include/deal.II/base/mpi.h:21,
4676:                  from /home/munch/sw-index/dealii/include/deal.II/distributed/tria_base.h:22,
4676:                  from /home/munch/sw-index/dealii/include/deal.II/distributed/fully_distributed_tria.h:22,
4676:                  from /home/munch/sw-index/dealii/tests/serialization/parallel_fullydistributed_construction_data_1.cc:19:
4676: /home/munch/sw-index/dealii/bundled/boost-1.70.0/include/boost/archive/detail/check.hpp: In instantiation of ‘void boost::archive::detail::check_object_level() [with T = unsigned int [0]]’:
4676: /home/munch/sw-index/dealii/bundled/boost-1.70.0/include/boost/archive/detail/iserializer.hpp:460:32:   required from ‘static void boost::archive::detail::load_non_pointer_type<Archive>::invoke(Archive&, T&) [with T = unsigned int [0]; Archive = boost::archive::tex$
_iarchive]’
4676: /home/munch/sw-index/dealii/bundled/boost-1.70.0/include/boost/archive/detail/iserializer.hpp:624:18:   required from ‘void boost::archive::load(Archive&, T&) [with Archive = boost::archive::text_iarchive; T = unsigned int [0]]’
4676: /home/munch/sw-index/dealii/bundled/boost-1.70.0/include/boost/archive/detail/common_iarchive.hpp:67:22:   required from ‘void boost::archive::detail::common_iarchive<Archive>::load_override(T&) [with T = unsigned int [0]; Archive = boost::archive::text_iarchive]’
4676: /home/munch/sw-index/dealii/bundled/boost-1.70.0/include/boost/archive/basic_text_iarchive.hpp:70:9:   required from ‘void boost::archive::basic_text_iarchive<Archive>::load_override(T&) [with T = unsigned int [0]; Archive = boost::archive::text_iarchive]’
4676: /home/munch/sw-index/dealii/bundled/boost-1.70.0/include/boost/archive/text_iarchive.hpp:82:52:   required from ‘void boost::archive::text_iarchive_impl<Archive>::load_override(T&) [with T = unsigned int [0]; Archive = boost::archive::text_iarchive]’
4676: /home/munch/sw-index/dealii/bundled/boost-1.70.0/include/boost/archive/detail/interface_iarchive.hpp:68:9:   [ skipping 116 instantiation contexts, use -ftemplate-backtrace-limit=0 to disable ]
4676: /home/munch/sw-index/dealii/bundled/boost-1.70.0/include/boost/archive/basic_text_iarchive.hpp:70:9:   required from ‘void boost::archive::basic_text_iarchive<Archive>::load_override(T&) [with T = dealii::parallel::fullydistributed::ConstructionData<2, 2>; Archive =
 boost::archive::text_iarchive]’
4676: /home/munch/sw-index/dealii/bundled/boost-1.70.0/include/boost/archive/text_iarchive.hpp:82:52:   required from ‘void boost::archive::text_iarchive_impl<Archive>::load_override(T&) [with T = dealii::parallel::fullydistributed::ConstructionData<2, 2>; Archive = boost
::archive::text_iarchive]’
4676: /home/munch/sw-index/dealii/bundled/boost-1.70.0/include/boost/archive/detail/interface_iarchive.hpp:68:9:   required from ‘Archive& boost::archive::detail::interface_iarchive<Archive>::operator>>(T&) [with T = dealii::parallel::fullydistributed::ConstructionData<2,
 2>; Archive = boost::archive::text_iarchive]’
4676: /home/munch/sw-index/dealii/tests/serialization/serialization.h:68:8:   required from ‘void verify(const T&, T&) [with T = dealii::parallel::fullydistributed::ConstructionData<2, 2>]’
4676: /home/munch/sw-index/dealii/tests/serialization/parallel_fullydistributed_construction_data_1.cc:51:9:   required from ‘void test(MPI_Comm) [with int dim = 2; MPI_Comm = ompi_communicator_t*]’
4676: /home/munch/sw-index/dealii/tests/serialization/parallel_fullydistributed_construction_data_1.cc:75:15:   required from here
4676: /home/munch/sw-index/dealii/bundled/boost-1.70.0/include/boost/archive/detail/check.hpp:60:5: error: static assertion failed: typex::value
4676:      BOOST_STATIC_ASSERT(typex::value);
4676:      ^~~~~~~~~~~~~~~~~~~
4676: In file included from /home/munch/sw-index/dealii/bundled/boost-1.70.0/include/boost/serialization/split_member.hpp:23,
4676:                  from /home/munch/sw-index/dealii/bundled/boost-1.70.0/include/boost/serialization/nvp.hpp:26,
4676:                  from /home/munch/sw-index/dealii/bundled/boost-1.70.0/include/boost/serialization/array_wrapper.hpp:19,
4676:                  from /home/munch/sw-index/dealii/bundled/boost-1.70.0/include/boost/archive/basic_binary_iprimitive.hpp:53,
4676:                  from /home/munch/sw-index/dealii/bundled/boost-1.70.0/include/boost/archive/binary_iarchive_impl.hpp:20,
4676:                  from /home/munch/sw-index/dealii/bundled/boost-1.70.0/include/boost/archive/binary_iarchive.hpp:20,
4676:                  from /home/munch/sw-index/dealii/include/deal.II/base/utilities.h:45,
4676:                  from /home/munch/sw-index/dealii/include/deal.II/base/tensor.h:27,
4676:                  from /home/munch/sw-index/dealii/include/deal.II/base/symmetric_tensor.h:23,
4676:                  from /home/munch/sw-index/dealii/include/deal.II/base/array_view.h:23,
4676:                  from /home/munch/sw-index/dealii/include/deal.II/base/mpi.h:21,
4676:                  from /home/munch/sw-index/dealii/include/deal.II/distributed/tria_base.h:22,
4676:                  from /home/munch/sw-index/dealii/include/deal.II/distributed/fully_distributed_tria.h:22,
4676:                  from /home/munch/sw-index/dealii/tests/serialization/parallel_fullydistributed_construction_data_1.cc:19:
4676: /home/munch/sw-index/dealii/bundled/boost-1.70.0/include/boost/serialization/access.hpp: In instantiation of ‘static void boost::serialization::access::serialize(Archive&, T&, unsigned int) [with Archive = boost::archive::text_iarchive; T = unsigned int [0]]’:
4676: /home/munch/sw-index/dealii/bundled/boost-1.70.0/include/boost/serialization/serialization.hpp:68:22:   required from ‘void boost::serialization::serialize(Archive&, T&, unsigned int) [with Archive = boost::archive::text_iarchive; T = unsigned int [0]]’
4676: /home/munch/sw-index/dealii/bundled/boost-1.70.0/include/boost/serialization/serialization.hpp:126:14:   required from ‘void boost::serialization::serialize_adl(Archive&, T&, unsigned int) [with Archive = boost::archive::text_iarchive; T = unsigned int [0]]’
4676: /home/munch/sw-index/dealii/bundled/boost-1.70.0/include/boost/archive/detail/iserializer.hpp:187:40:   required from ‘void boost::archive::detail::iserializer<Archive, T>::load_object_data(boost::archive::detail::basic_iarchive&, void*, unsigned int) const [with Ar
chive = boost::archive::text_iarchive; T = unsigned int [0]]’
4676: /home/munch/sw-index/dealii/bundled/boost-1.70.0/include/boost/archive/detail/iserializer.hpp:121:1:   required from ‘class boost::archive::detail::iserializer<boost::archive::text_iarchive, unsigned int [0]>’
4676: /home/munch/sw-index/dealii/bundled/boost-1.70.0/include/boost/archive/detail/iserializer.hpp:409:13:   required from ‘static void boost::archive::detail::load_non_pointer_type<Archive>::load_standard::invoke(Archive&, const T&) [with T = unsigned int [0]; Archive =
 boost::archive::text_iarchive]’
4676: /home/munch/sw-index/dealii/bundled/boost-1.70.0/include/boost/archive/detail/iserializer.hpp:422:38:   [ skipping 122 instantiation contexts, use -ftemplate-backtrace-limit=0 to disable ]
4676: /home/munch/sw-index/dealii/bundled/boost-1.70.0/include/boost/archive/basic_text_iarchive.hpp:70:9:   required from ‘void boost::archive::basic_text_iarchive<Archive>::load_override(T&) [with T = dealii::parallel::fullydistributed::ConstructionData<2, 2>; Archive =
 boost::archive::text_iarchive]’
4676: /home/munch/sw-index/dealii/bundled/boost-1.70.0/include/boost/archive/text_iarchive.hpp:82:52:   required from ‘void boost::archive::text_iarchive_impl<Archive>::load_override(T&) [with T = dealii::parallel::fullydistributed::ConstructionData<2, 2>; Archive = boost
::archive::text_iarchive]’
4676: /home/munch/sw-index/dealii/bundled/boost-1.70.0/include/boost/archive/detail/interface_iarchive.hpp:68:9:   required from ‘Archive& boost::archive::detail::interface_iarchive<Archive>::operator>>(T&) [with T = dealii::parallel::fullydistributed::ConstructionData<2,
 2>; Archive = boost::archive::text_iarchive]’
4676: /home/munch/sw-index/dealii/tests/serialization/serialization.h:68:8:   required from ‘void verify(const T&, T&) [with T = dealii::parallel::fullydistributed::ConstructionData<2, 2>]’
4676: /home/munch/sw-index/dealii/tests/serialization/parallel_fullydistributed_construction_data_1.cc:51:9:   required from ‘void test(MPI_Comm) [with int dim = 2; MPI_Comm = ompi_communicator_t*]’
4676: /home/munch/sw-index/dealii/tests/serialization/parallel_fullydistributed_construction_data_1.cc:75:15:   required from here
4676: /home/munch/sw-index/dealii/bundled/boost-1.70.0/include/boost/serialization/access.hpp:116:11: error: request for member ‘serialize’ in ‘t’, which is of non-class type ‘unsigned int [0]’
4676:          t.serialize(ar, file_version);
4676:          ~~^~~~~~~~~
4676: gmake[3]: *** [CMakeFiles/parallel_fullydistributed_construction_data_1.debug.dir/parallel_fullydistributed_construction_data_1.cc.o] Error 1
4676: gmake[2]: *** [CMakeFiles/parallel_fullydistributed_construction_data_1.debug.dir/all] Error 2
4676: gmake[1]: *** [CMakeFiles/parallel_fullydistributed_construction_data_1.debug.diff.dir/rule] Error 2
4676: gmake: *** [parallel_fullydistributed_construction_data_1.debug.diff] Error 2
4676: Test serialization/parallel_fullydistributed_construction_data_1.debug: BUILD
```

Reply:
Oh, but that's because you explicitly declared these to be length-zero arrays. Just leave them at length 1 as they were before the patch and this should compile fine. You're ignoring the contents of that array anyway, so who cares what length they have -- surely the extra four bytes don't break the bank ;-)

Reply:
@bangerth  The problem in that case is that those extra one (1D) and two (2D) integegers would have to be initialized with a certain value (what I do not really like), else the text version of the serialization (see tests/serialization/parallel_fullydistributed_construction_data_1.output) would produce random output, which makes comparison of the output of two test not possible...

Reply:
Right, so then just don't serialize these values (using the `if` statements I mentioned above) :-)

In the end, I don't care all that much. But experience shows that if you add a field to this class, you'll update two of the implementations you wrote and forget the third. It's easier to just have it all in one place.

By the way, you could also try to use `std::array<>` instead of the raw C-style arrays. I have no idea how boost decides to serialize a zero-sized `std::array`, though.

Reply:
https://en.cppreference.com/w/cpp/container/array states, however, that zero-sized `std::array` objects are perfectly legal.

Reply:
I have updated so that the arrays at least have the size of 1 and have added the two if-statements.